staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -70,6 +71,19 @@ type store struct {
 	pathPrefix    string
 	watcher       *watcher
 	pagingEnabled bool
+
+	// The following variables store the expiration time and id of the
+	// previous lease. If a new event has an expiration time earlier than
+	// the previous one, the previous lease will be reused to reduce
+	// etcd's overhead.
+	leaseLock               sync.Mutex
```

Comment:
This should really be abstracted into a sub type like “leaseManager”.

Reply:
+1 - that's a really good point

Reply:
done, working on unit test